Add new recommendation types (#132)
lmc-eu/matej-client-php: UPGRADE-4.0.md

> 

Reviewer: ```suggestion
```

# Upgrading from 3.x to 4.0

API client release 4.0 contains backward incompatible changes.

This guide will help you upgrade your codebase.

## Sorting renamed to ItemSorting

> Class `Sorting` was renamed to `ItemSorting`. All its methods and behaviour remained the same.

Reviewer: ```suggestion
Class `Sorting` was renamed to `ItemSorting`. All its methods and behavior remained the same.
```
(also for consistency, dtto as in README)

#### Before
```php
$response =  $matej->request()
    ->sorting(Sorting::create('user-id', ['item-id-1', 'item-id-2', 'item-id-3']))
    ->send();
```

#### After
```php
$response =  $matej->request()
    ->sorting(ItemSorting::create('user-id', ['item-id-1', 'item-id-2', 'item-id-3']))
    ->send();
```

## New types of recommendations
API client now supports new types of recommendation commands:
- Recommend items to user (`UserItemRecommendation`, originally `UserRecommendation`)
- Recommend users to user (`UserUserRecommendation`)
- Recommend items to item (`ItemItemRecommendation`)
- Recommend users to item (`ItemUserRecommendation`)

Each recommendation type supports different parameter options. Awailable parameters are

> described in table bellow:

Reviewer: ```suggestion
Each recommendation type supports different parameter options. Available parameters are
described in table below:
```

| hard_rotation |             `enableHardRotation`            |            ✅           |            ✅           |            ❌           |            ❌           |
| allow_seen    |                `setAllowSeen`               |            ✅           |            ❌           |            ❌           |            ✅           |
| min_relevance |            `setMinimalRelevance`            |  `ItemMinimalRelevance`  |            ❌           |            ❌           |  `UserMinimalRelevance`*  |
| filter        |            `addFilter` `setFilters`           |            ✅           |            ❌           |            ✅           |            ❌           |
| boost_rules   |             `addBoost` `setBoosts`            |            ✅           |            ❌           |            ✅           |            ❌           |
| model_name    |                `setModelName`               |            ✅           |            ✅           |            ✅           |            ✅           |
| properties    | `addResponseProperty` `setResponseProperties` |            ✅           |            ❌           |            ✅           |            ❌           |

> \* `UserMinimalRelevance` supports only `MEDIUM` and `HIGH` relevancies.

Reviewer: ```suggestion
\* `UserMinimalRelevance` supports only `MEDIUM` and `HIGH` relevances.
```

Also this part is missing in the readme (so I suggest moving it to readme only).

Each recommendation class provides a static constructor method `create` that accepts `userId` (recommendations for users) or `itemId` (recommendations for items).

## Recommending items to users
Class `UserRecommendation` was renamed to `UserItemRecommendation`. Class `MinimalRelevance` was
renamed to `ItemMinimalRelevance`. All other methods and attributes remained unchanged.

#### Before
```php
$recommendation = UserRecommendation::create('user-id', 'scenario')

>     ->setMinimalRelevance(MinimalRelevance::LOW())
>     ->setCount(5)
>     ->setRotationRate(1.0)
>     ->setRotationTime(3600);

Reviewer: I'd omit the last 3 lines - they are not relevant for the change, and keep just this:

```suggestion
    ->setMinimalRelevance(MinimalRelevance::LOW());
```

```

#### After
```php
$recommendation = UserItemRecommendation::create('user-id', 'scenario')
    ->setMinimalRelevance(ItemMinimalRelevance::LOW())
    ->setCount(5)
    ->setRotationRate(1.0)
    ->setRotationTime(3600);
```

## New recommendation types

> New types of recommendations can be created is similar way to `UserItemRecommendation`. For example:

Reviewer: ```suggestion
New types of recommendations can be created in a similar way to `UserItemRecommendation`. For example:
```

### Users that might be interested in an item
```php
$recommendation = ItemUserRecommendation::create('item-id', 'scenario')
    ->setCount(5)
    ->setAllowSeen(false);
```

### Users that are similar to a user
```php
$recommendation = UserUserRecommendation::create('user-id', 'scenario')
    ->setCount(5)
    ->setRotationRate(1.0)
    ->setRotationTime(3600);
```

### Items that are similar to an item
```php
$recommendation = ItemItemRecommendation::create('item-id', 'scenario')
    ->setCount(5);

> ```

Reviewer: Missing end of line on the end of file

# Upgrading from 3.x to 4.0

API client release 4.0 contains backward incompatible changes.

This guide will help you upgrade your codebase.

## New types of recommendations
API client now supports new types of recommendation commands:
- Recommend items to user (`UserItemRecommendation`, originally `UserRecommendation`)
- Recommend users to user (`UserUserRecommendation`)
- Recommend items to item (`ItemItemRecommendation`)
- Recommend users to item (`ItemUserRecommendation`)

> Each recommendation type supports different parameter options. Awailable parameters are

Reviewer: I'd rather link to the readme (you can link to specific #header), instead of duplicating this whole table below (it will be easier to maintain in case of bug, updates etc.) - users will be pointed to the most up-to-date documentation.
